src/main/java/org/opensearch/knn/index/memory/NativeMemoryCacheManager.java

```diff
@@ -0,0 +1,281 @@
+                maxWeight - getCacheWeightInKilobytes() - nativeMemoryEntryContext.calculateSizeInKb() <= 0
+        ) {
+            throw new NativeMemoryThrottleException("Failed to load \"" + nativeMemoryEntryContext.getKey() +
+                    "\" into memory.");
+        }
+
+        return cache.get(nativeMemoryEntryContext.getKey(), nativeMemoryEntryContext::load);
+    }
+
+    /**
+     * Invalidate entry from the cache.
+     *
+     * @param key Identifier of entry to invalidate
+     */
+    public void invalidate(String key) {
+        cache.invalidate(key);
+    }
+
+    /**
+     * Invalidate all entries in the cache.
+     */
+    public void invalidateAll() {
+        cache.invalidateAll();
+    }
+
+    /**
+     * Returns whether or not the capacity of the cache has been reached
+     *
+     * @return Boolean of whether cache limit has been reached
+     */
+    public Boolean isCacheCapacityReached() {
+        return cacheCapacityReached.get();
+    }
+
+    /**
+     * Sets cache capacity reached
+     *
+     * @param value Boolean value to set cache Capacity Reached to
+     */
+    public void setCacheCapacityReached(Boolean value) {
+        cacheCapacityReached.set(value);
+    }
+
+    /**
+     * Get the stats of all of the Elasticsearch indices currently loaded into the cache
+     *
+     * @return Map containing all of the Elasticsearch indices in the cache and their stats
+     */
+    public Map<String, Map<String, Object>> getIndicesCacheStats() {
+        Map<String, Map<String, Object>> statValues = new HashMap<>();
+        String indexName;
+        NativeMemoryAllocation.IndexAllocation indexAllocation;
+
+        for (Map.Entry<String, NativeMemoryAllocation> entry : cache.asMap().entrySet()) {
+
+            if (entry.getValue() instanceof NativeMemoryAllocation.IndexAllocation) {
+                indexAllocation = (NativeMemoryAllocation.IndexAllocation) entry.getValue();
+                indexName = indexAllocation.getOsIndexName();
+                statValues.putIfAbsent(indexName, new HashMap<>());
+
+                statValues.get(indexName).put(GRAPH_COUNT, ((Integer) statValues.get(indexName)
+                        .getOrDefault(GRAPH_COUNT, 0)) + 1);
+                statValues.get(indexName).putIfAbsent(StatNames.GRAPH_MEMORY_USAGE.getName(),
+                        getWeightForIndexInKilobytes(indexName));
+                statValues.get(indexName).putIfAbsent(StatNames.GRAPH_MEMORY_USAGE_PERCENTAGE.getName(),
+                        getWeightForIndexAsPercentage(indexName));
+            }
+        }
+
+        return statValues;
+    }
+
+
+    private void onRemoval(RemovalNotification<String, NativeMemoryAllocation> removalNotification) {
+        NativeMemoryAllocation nativeMemoryAllocation = removalNotification.getValue();
+        executor.execute(() -> {
+            nativeMemoryAllocation.writeLock();
+
+            try {
+                nativeMemoryAllocation.close();
```

Comment:
If close fails, it will most likely be caused by the jni free function failed. In this case, the program would most likely crash. Im not sure how I would handle this.

Reply:
so OpenSearch will crash?

Reply:
Yes, OpenSearch will crash if a nativeMemoryAllocation fails to close. This means it is unable to free memory it has allocated in the native layer.